include/libtorrent/packet_pool.hpp

```diff
@@ -78,78 +78,82 @@ namespace libtorrent
 		std::uint8_t buf[1];
 	};
 
+
+	struct packet_deleter
+	{
+		void operator()(packet* handle) const
+		{
+			std::free(handle);
+		}
+	};
+
+	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
+
+	template<std::size_t ALLOCATE_SIZE>
+	struct TORRENT_EXTRA_EXPORT packet_slab : private aux::vector<packet_ptr, std::size_t>
+	{
+		static const std::size_t allocate_size{ ALLOCATE_SIZE };
+
+		packet_slab(std::size_t limit) : m_limit(limit) { reserve(m_limit); }
+
+		void try_free(packet_ptr &p)
+		{
+			if (size() < m_limit)
+				push_back(std::move(p));
+		}
+
+		void try_alloc(packet_ptr &p)
+		{
+			if (empty())
+				p.reset(static_cast<packet*>(std::malloc(sizeof(packet) + allocate_size)));
+			else
+			{
+				p = std::move(back());
+				pop_back();
+			}
+		}
+	private:
+		std::size_t m_limit;
+	};
+
 	// single thread packet allocation packet pool
 	// can handle common cases of packet size by 3 pools
-	struct TORRENT_EXTRA_EXPORT packet_pool : single_threaded
+	struct TORRENT_EXTRA_EXPORT packet_pool : private single_threaded
 	{
 		packet *acquire(std::size_t allocate)
 		{
 			TORRENT_ASSERT(is_single_thread());
 			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
 
-			packet *p;
+			packet_ptr p;
 
-			if (allocate <= m_syn_slab.allocate_size) { p = m_syn_slab.try_alloc(); }
-			else if (allocate <= m_mtu_floor_slab.allocate_size) { p = m_mtu_floor_slab.try_alloc(); }
-			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { p = m_mtu_ceiling_slab.try_alloc(); }
-			else { p = static_cast<packet*>(malloc(sizeof(packet) + allocate)); }
+			if (allocate <= m_syn_slab.allocate_size) { m_syn_slab.try_alloc(p); }
+			else if (allocate <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_alloc(p); }
+			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_alloc(p); }
+			else { m_rest_slab.try_alloc(p); }
 
 			p->allocated = static_cast<std::uint16_t>(allocate);
```

Comment:
I think it would be a bit more robust if the ``allocated`` member was filled in by the ``malloc()`` function, no?
I guess that wouldn't quite work with the "rest" slab. it seems a bit redundant to implement the general case via a slab though. it holds state that's never used